#Lecture: CS412 Machine Learning | Homework 3
####Instructor: Alper Özpınar
####Author: Mert Atabey Dincer
####Student ID: 20637
####Email: atabeymert@sabanciuniv.edu
####Date: 04.08.2020



In [59]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [60]:
url="https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

columnNames = ["SepalLength", "SepalWidth", "PetalLength", "PetalWidth", "Output"]
df = pd.read_csv(url, index_col=None, header=None, names=columnNames) #Reads dataset from url



In [61]:
attributeCols = ["SepalLength", "SepalWidth", "PetalLength", "PetalWidth"]

X = df[attributeCols] #Seperate features from labels.

y = df.Output

In [62]:
#Split dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state = 7) #75% training and 25% test

In [63]:
from sklearn.metrics import  accuracy_score, f1_score, precision_score, recall_score, classification_report

depths = [50,20,15,10,5,3]
tree_index = 0 

#To be used to identify best tree
best_tree_index = 0
best_accuracy = 0
best_f1 = 0
best_impurty = ""
best_depth = 0

#To be used to identify worst tree
worst_tree_index = 0
worst_accuracy = 1
worst_f1 = 1
worst_impurty = ""
worst_depth = 0

for i in depths:
  tree_index = tree_index+1
  dtc_GINI = DecisionTreeClassifier(criterion="gini", max_depth=i) #Create a decision tree classifier (dtc)
  dtc_GINI = dtc_GINI.fit(X_train, y_train) #Train the decision tree
  prediction_GINI = dtc_GINI.predict(X_test)

  accuracy_gini = accuracy_score(y_test, prediction_GINI)
  f1_gini = f1_score(y_test, prediction_GINI, average='weighted')
  precision_gini = precision_score(y_test, prediction_GINI, average='weighted')
  recall_gini = recall_score(y_test, prediction_GINI, average='weighted')

  if accuracy_gini > best_accuracy and f1_gini > best_f1:
    best_tree_index = tree_index
    best_accuracy = accuracy_gini
    best_f1 = f1_gini
    best_precision = precision_gini
    best_recall = recall_gini
    best_impurty = "GINI"
    best_depth = i
  elif  accuracy_gini < worst_accuracy and f1_gini < worst_f1:
    worst_tree_index = tree_index
    worst_accuracy = accuracy_gini
    worst_f1 = f1_gini
    worst_precision = precision_gini
    worst_recall = recall_gini
    worst_impurty = "GINI"
    worst_depth = i

###########################################################################################

  tree_index = tree_index+1
  dtc_Entropy = DecisionTreeClassifier(criterion="entropy", max_depth=i)
  dtc_Entropy = dtc_Entropy.fit(X_train, y_train) 
  prediction_Entropy = dtc_Entropy.predict(X_test)

  accuracy_entropy = accuracy_score(y_test, prediction_Entropy)
  f1_entropy = f1_score(y_test, prediction_Entropy, average='weighted')
  precision_entropy = precision_score(y_test, prediction_Entropy, average='weighted')
  recall_entropy = recall_score(y_test, prediction_Entropy, average='weighted')

  if accuracy_gini > best_accuracy and f1_gini > best_f1:
    best_tree_index = tree_index
    best_accuracy = accuracy_entropy
    best_f1 = f1_entropy
    best_impurty = "Entropy"
    best_depth = i
  elif  accuracy_gini < worst_accuracy and f1_gini < worst_f1:
    worst_tree_index = tree_index
    worst_accuracy = accuracy_entropy
    worst_f1 = f1_entropy
    worst_impurty = "Entropy"
    worst_depth = i
  

print("The best tree is the {}. created tree which is composed of:\n Accuracy: {:.3f}\n F1 Score: {:.3f}\n Impurtiy Type: {}\n Depth: {} \n".format(best_tree_index,best_accuracy,best_f1,best_impurty,best_depth))
print("The worst tree is the {}. created tree which is composed of:\n Accuracy: {:.3f}\n F1 Score: {:.3f}\n Impurtiy Type: {}\n Depth: {}".format(worst_tree_index,worst_accuracy,worst_f1,worst_impurty,worst_depth))


The best tree is the 1. created tree which is composed of:
 Accuracy: 0.921
 F1 Score: 0.921
 Impurtiy Type: GINI
 Depth: 50 

The worst tree is the 7. created tree which is composed of:
 Accuracy: 0.895
 F1 Score: 0.895
 Impurtiy Type: GINI
 Depth: 10


#Ran entire code 3 times
---
##Result 1: Between GINI and Entropy there is not much performance difference. We can say they are about the same. One of them is not better than the other. This case depends only on the data.

####The best tree is the 1. created tree which is composed of:
* Accuracy: 0.947
* F1 Score: 0.947
* Impurtiy Type: GINI
* Depth: 50 

####The worst tree is the 2. created tree which is composed of:
* Accuracy: 0.921
* F1 Score: 0.921
* Impurtiy Type: Entropy
* Depth: 50
---
##Result 2: Having too little depth can cause underfitting which decreases performance of learner in terms of prediction.

####The best tree is the 5. created tree which is composed of:
* Accuracy: 0.947
* F1 Score: 0.947
* Impurtiy Type: GINI
* Depth: 15 

#####The worst tree is the 11. created tree which is composed of:
* Accuracy: 0.895
* F1 Score: 0.895
* Impurtiy Type: GINI
* Depth: 3
---
##Result 3:In contrast, having too much depth can also cause overfitting which get us away from generalization approach that results with performance loss.

####The best tree is the 3. created tree which is composed of:
* Accuracy: 0.947
* F1 Score: 0.947
* Impurtiy Type: GINI
* Depth: 20 

####The worst tree is the 2. created tree which is composed of:
* Accuracy: 0.921
* F1 Score: 0.921
* Impurtiy Type: Entropy
* Depth: 50
